editor/config_loader.hpp

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include "base/exception.hpp"
+#include "base/logging.hpp"
+
+#include "std/chrono.hpp"
+#include "std/mutex.hpp"
+#include "std/string.hpp"
+#include "std/thread.hpp"
+
+namespace pugi
+{
+class xml_document;
+}
+
+namespace editor
+{
+// Class for multithreaded interruptable waiting.
+class Waiter
+{
+public:
+  template <typename Rep, typename Period>
+  bool Wait(duration<Rep, Period> const & d)
+  {
+    unique_lock<mutex> lock(m_mutex);
+
+    if (m_interrupted)
+      return false;
+
+    try
+    {
+      m_event.wait_for(lock, d, [this]() { return m_interrupted; });
+    }
+    catch (RootException const & ex)
+    {
+      LOG(LERROR, (ex.Msg()));
+    }
+
+    return true;
+  }
+
+  void Interrupt();
+
+private:
+  bool m_interrupted = false;
+  mutex m_mutex;
+  condition_variable m_event;
+};
+
+DECLARE_EXCEPTION(ConfigLoadError, RootException);
+
+// Class which loads config from local drive, checks hash
+// for config on server and downloads new config if needed.
+class ConfigLoader
+{
+public:
+  ConfigLoader(pugi::xml_document & doc, mutex & mtx);
```

Comment:
I generally dislike the idea to pass mutex by reference. It's much better to create a thread-safe wrapper around xml_document to guard parts of it.